### ISSUE TITLE
L1 non-fill proof

### ISSUE THREAD

**Issue by cducrest** [beamer-bridge/beamer#346]
We had this discussion with Fred about the following problem:

> What happens if `Charles` self-challenges a false claim on `Alice's` request that cannot be filled (e.g. she request a transfer for an unsupported token / too high a value / off-chain sequencer ....). Then no correct claim will be able to terminate before `Charles's` claim and he will be able to withdraw `Alice's` deposit.

A potential solution is to have some non-fill proof submitted from destination roll-up to source roll-up via L1 proving that no fill for a corresponding request hash was done. Here's how it would look:

1) Alice sees Charles’ false claim. Observers knows that the fill ID is false, and the the request is not filled by anyone else.
2) Bob triggers an L2 -> L1 message which states that the request is not filled yet. This needs to be triggered `finalization_time` of target roll-up before Charles' claim end.
3) The non-fill proof submitted to source roll-up should enable Bob to earn the deposit of Charles, to incentivize Bob's participation.
4) Charles should not be able to withdraw its stake or Alice's deposit.
5) Alice can withdraw her deposit if the `expiry` time on her request ended.

Bob's intervention is only necessary if Charles challenged and counter-challenged himself, meaning that the result would be that his claim over Alice's deposit is successful. This means that we need to enforce that the counter-challenge stake should be sufficient to incentivize Bob's participation to submit the L1 non-fill proof.

We need to either find a better idea or implement it.

**Comment by cducrest** [comment]
Either we allow non-fill proofs to be submitted at any time, and we don't add the `valid until block` field to the request hash. The proof will need to have the block number at which it was submitted, and it will need to be registered in the resolution registry for the request manager to check whether the non-fill proof was submitted too early.

Or non-fill proofs cannot be submitted before `valid until block` as verified in the fill manager, so the `valid until block` must be part of the request hash (and the fill call).

Second solution seems easier to implement, you are right (albeit more gas intensive in the optimist case, but I don't think we care).

**Comment by fredo** [comment]
Is it really more gas efficient? 

I understand that we must add the `valid until block` in 1) but at the same time we would need to add the `filled block number` upon any claim. This would be necessary in order for an honest claim not to be blocked by the non-fill-proof message. Furthermore we would need to pass the non-fill block number via L1 resolution. 

Whereas in 2) we have the implicit block number of the non -fill proof which is `valid until block`. We know that once a non-fill proof message is sent, it happened after `valid until block` but we also know it cannot be filled anymore afterwards. So all claims are invalid.

The "downside" of 2) is that we introduce new timing constraints for request expiries. But I also double checked with the Business Team, that having this constraint is not a problem. So basically, reducing maximum request expires down to 1 hour or 100 blocks or something like that is even desirable.

**Comment by fredo** [comment]
I realized that also in the current implementation the maximum request expiry should be `challengePeriodExtension`.

It's the same trick.

1) Charlie claims Non-filled request
2) Charlie challenges own claim
3) Charlie waits for 7 days (leaving 1 day (or challengePeriodExtension) in the challengePeriod)
4) in the mean time but after `challengePeriodExtension` time, Bob fills the request
5) Bob claims
6) Charlie challenges Bob's claim
7) Before 3) is about to end, charlie counter-challenges, ChallengePeriod does not get extended
8) After 1 day, Charlies claim ends and he can withdraw the funds directly because the fill message is still in-flight

**Comment by fredo** [comment]
The fundamental problem with actions happening after `challengePeriodExtension` remains the same. While we can forbid fills with the expiry, we cannot with Non-fill proofs. They must happen before the time constraint.
Currently we cannot incentivize a Non-fill-proof sender. As he is not part of any challenge game.

There are two fundamental problems: 

a) After challengePeriodExtension it is not safe to do anything on the target rollup
b) There is no incentivization for a non-participant of a challenge game to call Non-Fill



The following Options we have:

1)
Allow Non-Fill Claim. He will get reimbursed by the user. Note, that the user will pay a fee although the request expired. Here the Non-fill proof still must happen before `challengePeriodExtension` of the false claim

fixes b) by the user paying the watcher, does not fix a)

2)
Step in the challenge as the Non-Fill Prover. Allow any participant to step in a challenge once by stating that they are the one who called Non-Fill Proof. This will extend the challengePeriod to another `finalization_time + challengePeriodExtension`. Afterwards funds need to be distributed correctly between the three parties.
The problem is that as long as the challenger is leading, we cannot differentiate if the behavior of the challenger is correct. Once the claimer counter challenges we must assume that this is a self-challenge attack and call Non-fill proof. Thus the claimer needs to provide enough funds to go to L1, so that there is enough stake in the game to step in as the Non-fill prover.

fixes a) by extending a challenge, fixes b) by redistribution of challenge. Could change game theoretical incentivization of the challenge game. We need to think about that.

**Comment by fredo** [comment]
Some fundamental thought: 

We set the claim period as such, that at least one honest participant can intervene. If not the protocol is not safe.
Ideally, we make the same assumption for the Non-fill prover. If there is enough time after request expiry AND there is a claim, any participant should intervene by sending the non-fill proof on the target L2. This would solve the problem of the timing issue of a). 
Open question which remains: How is this action incentivized?

**Comment by fredo** [comment]
I think we can relax the time constraint of `maxValidityPeriod` and `challengePeriodExtension` by adding the following tweak: 

```
initialChallengePeriod = finalization_time + ChallengePeriodExtension + RequestMaxValidity
```

This would give the user the time to react after the request is expired as laid out in the rest of the protocol. (i.e. `challengePeriodExtension` or `claimPeriod`)

**Comment by fredo** [comment]
I just realized that `validUntilBlock` cannot be verified as the RequestManager has no information about the current block of the target rollup. But what we can do is:

- limit the validUntil to `maxValidityPeriod`
- only allow fills on the target rollup where `validUntilBlock < block.number + maxValidUntilBlock`. 
- It’s up to the agent to reason about whether it makes sense to fill this request or not. It is effectively limited to ~validUntil

**Comment by fredo** [comment]
I have another proposal: 

After thinking about all of this I think we must prevent the self challenge issue on a more profound level. Instead of adding edge case logic for the Non-Fill Case, we must prevent that charlie can control the outcome of a challenge.
The way to do this is to allow multiple parties to join the challenge. We have talked about this before, and decided against it because the original challenger still must be rewarded for stepping in in the first place. Instead of replacing the original challenger by the new challenger we must make them equally stakeholders in the challenge game. That's the crucial tweak to the multi party challenge scheme.

This is done by a bookkeeping mechanism of who bid what. We need to keep who bid which amount in total and who was the last won who bid. Let's make an example:

1) charlie claims 5

| claimer | challenger |
|-----------|-----------------|
| 5 | |

2) Bob challenge 6

| claimer | challenger |
|-----------|-----------------|
| 5 | Bob(6) |

3) charlie counter challenges 5

| claimer | challenger |
|-----------|-----------------|
| 10 | Bob(6) |

4) David challenge 5

| claimer | challenger |
|-----------|-----------------|
| 10 | Bob(6), David(5,L) |

4) Charlie counter challenge 3

| claimer | challenger |
|-----------|-----------------|
| 13 | Bob(6), David(5,L) |

5) Bob challenge 3

| claimer | challenger |
|-----------|-----------------|
| 13 | Bob(9, L), David(5) |

6) Challenge ends

Payout scheme of Charlies 13 tokens

David -> 5
Bob -> 8 (remaining tokens)


### Rationale of book keeping
the rationale behind the book keeping is that anybody who entered into a challenge and put in some stake, should be rewarded when the his position was successful by the amount he put in. Only the last one who made the last bid will get the remaining tokens as there is not enough stake on the other side. 
Actually this kind of bookkeeping is still true if the number of `bidders = 1`. The last one gets the remaining tokens.

This allows honest watchers to enter into any challenge at any point in time provoking the dishonest counterpart to either bid more (and thus lose more) or to end the challenge game. The potential maximum gain for one for each bid is `max_gain = stake_winning_party - stake_losing_party`.

One special case where the dishonest party was leading, but the L1 resolution resolved otherwise, we would have excess stake left. I would pay it out to the one who executed the L1 transaction as an additional incentive to do so.

### Non-Fill Proofs
This mechanism would not be needed for filled requests, as we can assume there will be the true filler who has an interest of actually withdrawing the funds before the dishonest challenger can end his game. This is different for unfilled requests. So this mechanism is needed here.

We will have Non-Fill Proof messages. I believe that Non-Filled Proof messages must be able to be sent at any point in time. `ValidUntilBlock` cannot be validated and thus can have any value. So it could be far in the future. We should discuss the necissity of `validUntilBlock` in its own issue.

The claimer will have to pass `fill_blocknumber` and `fill_id` to the claim. Any honest watcher must trigger the Non-Fill Proof message on target rollup to ensure message arrival before the challenge ends.
If Charlie makes himself the leading party in the challenge. David can step in and counter challenge. He will get now `Claimer_stake - first_challenger_stake` for sure if the challenge ends like that. If Charlie counter challenges (as the claimer) David will get his own stake in return if the challenge is successful. The normal challenge game continues until it either ends or someone triggered L1. If david's stake is enough to go to L1, he will do it. Since the is a Non-Fill proof message ready, he can do it.

Additionally, in the future we could pass the executing addresses of the Non-fill proof and L1 resolution and pay the one who executed the transaction.

I think this scheme resolves the problem to raise funds from dishonest participants to go to L1. It currently does not 100% resolve the reimbursement of the L2 transaction for the Non-Fill message. But I think we can ignore that for now and solve this later.

**Comment by fredo** [comment]
Results from discussion with @cducrest:
 
 - We need to do multiparty bidding on challenger side to prevent self-challenge blocking
 - If excess capital is in the game, the last bidder will win the exceeding capital
     - This will incentivize outbidding the dishonest winner
     - This will incentivize the last bidder to call L1
 - After the initial claim stakes are set, we increase the minimum bid to some value as high to be sufficient to go to L1. Otherwise, a new participant might be victim to a griefing attack on the opportunity cost of his stake and thus not being incentivized to enter into the game in the first place.

**Comment by fredo** [comment]
@cducrest 

Following idea: 

Currently we store the following mapping to prevent multiple fills for the same request:

```
mapping(bytes32 => bool) public fills;
```

We could change it to: 

```
mapping(bytes32 => bytes32) public fills;
```

where the value of the mapping is the fill hash described as `fill_hash = abi.encodePacked(request_hash, block.number, fill_id)`.

We then create a function to send `invalid_fill_hash` messages to L1.

PSEUDO

```sol
function sendInvalidFillHash(requestHash, blockNumber, fillId) {
    fillHash = _createFillHash(requestHash, blockNumber, fillId);
    if(blockNumber > block.number){
        sendInvalidMessage(fillHash);
    }
    if(fills[requestHash] != fillHash){
        sendInvalidMessage(fillHash);
    }
}
```

For sure I have forgotten some edge cases... but I hope you get the point.
The underlying assumption says: There is only one valid fill hash per request. And if there is none yet then nobody is able to guess what the correct fill hash will be. So we can invalidate any fill hash because it is just guessing. 
This would allow us to use the same function for the attacker who 
a) chooses the block number to be far in the future
b) chooses a past block number but wrong fill id

**Comment by cducrest** [comment]
Sounds good
